### DISCUSSION TITLE
Adding independent servo to CF Bolt 1.1

### DISCUSSION THREAD

**Discussion by vBruegge** [bitcraze/discussions#988] [Q&A]
Hi,
I'm working on a custom deck for the CF Bolt 1.1 which adds a new servo to the drone. It's completely independent of all controls and just used for actuating a grabber. I looked through the firmware and couldn't find an option to add the servo without changing the complete ```motor.c``` and ```motor_def.c``` files. Is that the only possible way to do this and/or how can it exactly be achieved?
Thanks in advance!

**Comment by knmcguire** [comment]
Hi! So the flapper drone has 2 servo's and 2 motors. You can check out the power distribution here : https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/power_distribution_flapper.c. Or is this an extra servo on top of the 4 motors for a quadcopter?

I would imagine that you would need some kind of motor driver in between though to handle the PWM values.

**Reply by vBruegge** [reply]
Hi, it's an extra servo on top of the other 4 motors. So I want to control 5 servos in total, 4 for flying the drone and one to actuate the grabber.

**Reply by tobbeanton** [reply]
There is no easy way to do this (I think). You would have to add another motor to the motor driver and implement a servo setting. This probably implies changing a lot of code since a lot is assumed to be 4 motors. Another way to do it, probably easier, would be to make a new servo driver since you would have to use one of the timers on the deck connector such as the:
```
//Deck IO1, PB8, TIM4_CH3
//Deck IO2, PB5, TIM3_CH2
```
Could you perhaps add a issue for this in the crazyflie-firmware repo?

**Reply by vBruegge** [reply]
There is already an issue (#1024) regarding this topic, but as far as I can tell, there was no further development until now. Since everything is hardcoded for 4 motors, the only option I can think of is adding a new file containing the motor drive code of the ```motor.c ``` and adjust it for using one servo. Do you have any suggestions or ideas how I can tackle this problem in another way?

**Reply by tobbeanton** [reply]
The "correct" way I think would be to add support in motors.c. Hard to say how difficult it will be without trying to do it. There is the define NBR_OF_MOTORS which is used in many places but I'm pretty sure there are a lot of code that doesn't support just changing the define. Also servo support would have to be added so a neutral position could be selected.

Since the servo will not be used for stabilization maybe a separate file would make more sense.

**Reply by matejkarasek** [reply]
I stumbled across this answer only now...

But there are two options indeed:
1) separate servo driver like in #1005 
Since this is a functionality we often need, I am currently trying to make a version, which we could push into master...

2) enabling more motors in the power distribution
This gets trickier indeed. 
I made this work with an older code which allowed up to 8 motors in the power distribution:
https://github.com/flapper-drones/crazyflie-firmware/tree/more_motors
We are using this to have an extra servo on the flapper

Unfortunately, the motor-related code has changed substantially shortly after I made this and I wasn't able to merge my additions after these new changes... An attempt that does not work (yet) is here:
https://github.com/flapper-drones/crazyflie-firmware/tree/more_motors_new

**Reply by matejkarasek** [reply]
Here the PR for option 1) : https://github.com/bitcraze/crazyflie-firmware/pull/1340

**Reply by tobbeanton** [reply]
Thanx! I'll have a look at it as soon as I find some time!

**Reply by matejkarasek** [reply]
For anyone still looking for this functionality, it is now in master: https://github.com/bitcraze/crazyflie-firmware/pull/1340
